backend/src/middleware/handleHtmlContent/handleContentData.js

```diff
@@ -1,39 +1,57 @@
 import extractMentionedUsers from './notifications/extractMentionedUsers'
 import extractHashtags from './hashtags/extractHashtags'
 
-const notify = async (postId, idsOfMentionedUsers, context) => {
+const notifyMentions = async (label, id, idsOfMentionedUsers, context) => {
+  if (!idsOfMentionedUsers.length) return
+
   const session = context.driver.session()
   const createdAt = new Date().toISOString()
-  const cypher = `
-    MATCH(p:Post {id: $postId})<-[:WROTE]-(author:User)
-    MATCH(u:User)
-    WHERE u.id in $idsOfMentionedUsers
-    AND NOT (u)<-[:BLOCKED]-(author)
-    CREATE(n:Notification{id: apoc.create.uuid(), read: false, createdAt: $createdAt})
-    MERGE (p)-[:NOTIFIED]->(n)-[:NOTIFIED]->(u)
+  let cypher
+  if (label === 'Post') {
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-15T23:23:34Z" title="Friday, August 16th 2019, 1:23:34 am +02:00">Aug 16, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,39 +1,46 @@
 import extractMentionedUsers from './notifications/extractMentionedUsers'
 import extractHashtags from './hashtags/extractHashtags'
 
-const notify = async (postId, idsOfMentionedUsers, context) => {
+const notifyMentions = async (label, id, idsOfMentionedUsers, context) => {
+  if (!idsOfMentionedUsers.length) return
+
   const session = context.driver.session()
   const createdAt = new Date().toISOString()
   const cypher = `
-    MATCH(p:Post {id: $postId})<-[:WROTE]-(author:User)
-    MATCH(u:User)
+    MATCH (source)
+    WHERE source.id = $id AND $label IN LABELS(source)
+    MATCH (source)<-[:WROTE]-(author: User)
+    MATCH (u: User)
     WHERE u.id in $idsOfMentionedUsers
     AND NOT (u)<-[:BLOCKED]-(author)
```


The `author` must be different here. Consider the following situation:

1. (personA)-[:BLOCKED]->(personB)
2. (personA)-[:WROTE]->(post)
3. (personC)-[:WROTE]->(comment)-[:COMMENTED]->(post)

Now what happens if `personC` would mention `personB` in post authored by `personA`? As `personB` you would get a notification linked to a post and following it you get a 404. **That's** the situation we want to avoid.

It is no problem right now if the author of a comment mentions a blocked user. The blocked user could still follow the link and not get a 404. We would change this behaviour as soon as your blocked users cannot see your comments anymore. But I don't think that we will implement this behaviour, because you would have totally different conversation flows per user account.

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-08-15T23:29:16Z" title="Friday, August 16th 2019, 1:29:16 am +02:00">Aug 16, 2019</time>_

---


Could you write a backend test for this? Have a look into [this example here](https://github.com/Human-Connection/Human-Connection/blob/master/backend/src/middleware/handleHtmlContent/handleContentData.spec.js#L170).

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-16T16:09:11Z" title="Friday, August 16th 2019, 6:09:11 pm +02:00">Aug 16, 2019</time>_

---


Done 😆